File: 3월1주차/leetcode-108-Convert Sorted Array to Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def bfs(array, node, lr): # lr이 0이면 node의 왼쪽, 1이면 오른쪽에 붙인다
            if len(array) == 0:
                return
            l = len(array)
            l = l//2
            if lr == 0:
                node.left = TreeNode(array[l])
                bfs(array[:l], node.left, 0)
                bfs(array[l+1:], node.left, 1)
            else:
                node.right = TreeNode(array[l])
                bfs(array[:l], node.right, 0)
                bfs(array[l+1:], node.right, 1)
        root = TreeNode(0)
        bfs(nums, root, 0)
        return root.left

class Solution: # 솔루션을 참고한 좀 더 깔끔한 재귀 코드
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        if len(nums) == 0:
            return None
        l = len(nums) // 2
        node = TreeNode(nums[l])
        node.left = self.sortedArrayToBST(nums[:l])
        node.right = self.sortedArrayToBST(nums[l+1:])
        return node

File: 5월 1주차/leetcode-108-Convert Sorted Array to Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if (nums.size() == 0) return nullptr;
        int n = nums.size();
        TreeNode *nodeptr = new TreeNode(nums[n/2]);
        vector<int> left(nums.begin(), nums.begin() + n/2);
        vector<int> right(nums.begin() + n/2 + 1, nums.end()); 
        nodeptr->left = sortedArrayToBST(left);
        nodeptr->right = sortedArrayToBST(right);
        return nodeptr;
    }
};